packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol

```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+    }
+
+    /// @dev IConstantFlowAgreementV1.increaseFlowAllowanceWithPermissions implementation
+    function increaseFlowAllowanceWithPermissions(
+        ISuperfluidToken token,
+        address flowOperator,
+        uint8 permissionsToAdd,
+        int96 addedFlowRateAllowance,
+        bytes calldata ctx
+    ) external override returns (bytes memory newCtx) {
+        newCtx = ctx;
+        ISuperfluid.Context memory currentContext =
+            _validateUpdateFlowOperatorDataInput(token, flowOperator, permissionsToAdd, addedFlowRateAllowance, ctx);
+
+        (bytes32 flowOperatorId, uint8 oldPermissions, int96 oldFlowRateAllowance) =
+            getFlowOperatorData(token, currentContext.msgSender, flowOperator);
+
+        FlowOperatorData memory flowOperatorData;
+        flowOperatorData.permissions = deltaAddPermissions(oldPermissions, permissionsToAdd);
+        // @note this will revert if it overflows
+        int96 newFlowRateAllowance = oldFlowRateAllowance + addedFlowRateAllowance;
+        flowOperatorData.flowRateAllowance = newFlowRateAllowance;
+        token.updateAgreementData(flowOperatorId, _encodeFlowOperatorData(flowOperatorData));
+
+        emit FlowOperatorUpdated(
+            token, currentContext.msgSender, flowOperator, flowOperatorData.permissions, newFlowRateAllowance
+        );
+    }
+
+    /// @dev IConstantFlowAgreementV1.decreaseFlowAllowanceWithPermissions implementation
+    function decreaseFlowAllowanceWithPermissions(
+        ISuperfluidToken token,
+        address flowOperator,
+        uint8 permissionsToRemove,
+        int96 subtractedFlowRateAllowance,
+        bytes calldata ctx
+    ) external override returns (bytes memory newCtx) {
+        newCtx = ctx;
+        ISuperfluid.Context memory currentContext = _validateUpdateFlowOperatorDataInput(
+            token, flowOperator, permissionsToRemove, subtractedFlowRateAllowance, ctx
+        );
+
+        (bytes32 flowOperatorId, uint8 oldPermissions, int96 oldFlowRateAllowance) =
+            getFlowOperatorData(token, currentContext.msgSender, flowOperator);
+
+        FlowOperatorData memory flowOperatorData;
+        flowOperatorData.permissions = deltaRemovePermissions(oldPermissions, permissionsToRemove);
+        int96 newFlowRateAllowance = oldFlowRateAllowance - subtractedFlowRateAllowance;
+
+        // @note this defends against negative allowance
+        if (newFlowRateAllowance < 0) revert CFA_ACL_NO_NEGATIVE_ALLOWANCE();
+        flowOperatorData.flowRateAllowance = newFlowRateAllowance;
+
+        token.updateAgreementData(flowOperatorId, _encodeFlowOperatorData(flowOperatorData));
+
+        emit FlowOperatorUpdated(
+            token, currentContext.msgSender, flowOperator, flowOperatorData.permissions, newFlowRateAllowance
+        );
+    }
+
+    function deltaAddPermissions(uint8 existingPermissions, uint8 permissionDelta)
+        public
```

Comment:
What are the heuristics to decide whether to make this little helper public in the ABI or not?

Reply:
This is public mainly for ease of testability, but given the earlier comment regarding testing the implementation with implementation, the value may lie there in not making it public.

However, testing bitwise operations with fuzzing seems a bit pedantic here.